proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,343 @@
+---
+simd: '0167'
+title: Loader-v4
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Review
+created: 2024-08-15
+feature: 2aQJYqER2aKyb3cZw22v4SL2xMX7vwXBRWfvS4pTrtED
+---
+
+## Summary
+
+A new upgradeable loader which only requires a single account per program.
+
+## Motivation
+
+Loader-v3, which is currently the only deployable loader, requires two accounts
+per program. And has three account types: Program, program-data and buffer.
+This was a workaround to circumvent the finality of the `is_executable` flag,
+which will be ignored by the program runtime from SIMD-0162 onwards.
+Consequentially, this setup of the program account being a proxy account,
+containing the address of the actual program data account, is no longer
+necessary and should be removed. Likewise the distinction of executable program
+accounts and non-executable buffer accounts is also no longer necessary and
+should be removed as well.
+
+In loader-v3 every instruction which modified the program data had to re-verify
+the ELF in the end. Instead we are now aiming for a more modular workflow which
+explicates these steps in a Retract-Modify-Redeploy sequence of instructions
+that allows multiple modifications to share one verification of the ELF in the
+end. This sequence can be a single transaction or split across multiple
+transactions.
+
+Another issue with loader-v3 is that the executable file stored in the
+programdata account is misaligned relative to the beginning of the account.
+This currently requires a copy in the ELF loader to re-align the program.
+
+Additionally, there currently is no complete specification of the loaders
+program management instructions. This proposal would thus fill that gap once
+loader-v4 goes into production.
+
+See impact for further motivation.
+
+## Alternatives Considered
+
+A delay-visibility-free redeployment could be achieved by keeping the swap
+program around until the end of the slot. This would however mean that two
+accounts per program must be loaded until the dapp developer reclaims the
+second one. That would defy the purpose of this proposal which is to get rid
+of the proxy account.
+
+## New Terminology
+
+None.
+
+## Detailed Design
```

Comment:
I don't see any mention about loader-v4 instructions being prohibited from being invoked via cpi so this would be the first time we allow setting the `is_executable` flag on an account in a CPI program execution. If an sbf program invoked a cpi deploy, it wouldn't be able to see that flag changing because we currently don't update the `is_executable` flag for the sbf caller program. We should either prohibit any instructions that change the `is_executable` flag from being cpi's or we should update the SVM to copy the value back into the caller's memory.